service/src/java/org/apache/hive/service/auth/ldap/LdapGroupCallbackHandler.java

```diff
@@ -0,0 +1,135 @@
+import javax.security.sasl.AuthenticationException;
+import javax.security.sasl.AuthorizeCallback;
+import java.io.IOException;
+
+/**
+ * Callback handler that enforces LDAP filters on Kerberos-authenticated users.
+ * This handler applies the same LDAP filter resolution used for LDAP authentication
+ * to Kerberos users, ensuring consistent authorization policies.
+ */
+public class LdapGroupCallbackHandler implements CallbackHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(LdapGroupCallbackHandler.class);
+
+  private final HiveConf conf;
+  private final boolean enableLdapGroupCheck;
+  private final CallbackHandler delegateHandler;
+  private final DirSearchFactory dirSearchFactory;
+  private final Filter filter;
+
+  private final KerberosLdapFilterEnforcer filterEnforcer = KerberosLdapFilterEnforcer.INSTANCE;
+
+  public LdapGroupCallbackHandler(HiveConf conf) {
+    this(conf, new LdapSearchFactory(), new SaslRpcServer.SaslGssCallbackHandler());
+  }
+
+  @VisibleForTesting
+  LdapGroupCallbackHandler(HiveConf conf, DirSearchFactory dirSearchFactory, CallbackHandler delegateHandler) {
+    this.conf = conf;
+    this.delegateHandler = delegateHandler;
+    this.dirSearchFactory = dirSearchFactory;
+    this.enableLdapGroupCheck = conf.getBoolVar(HiveConf.ConfVars.HIVE_SERVER2_LDAP_ENABLE_GROUP_CHECK_AFTER_KERBEROS);
+    this.filter = enableLdapGroupCheck ? LdapAuthenticationProviderImpl.resolveFilter(conf) : null;
+
+    if (enableLdapGroupCheck && filter == null) {
+      LOG.warn("LDAP group check enabled but no filters configured");
+    }
+  }
+
+  @VisibleForTesting
+  public static LdapGroupCallbackHandler createForTesting(HiveConf conf, DirSearchFactory dirSearchFactory,
+      CallbackHandler delegateHandler) {
+    return new LdapGroupCallbackHandler(conf, dirSearchFactory, delegateHandler);
+  }
+
+  @Override
+  public void handle(Callback[] callbacks) throws IOException, UnsupportedCallbackException {
+    delegateHandler.handle(callbacks);
+
+    for (Callback callback : callbacks) {
+      if (!(callback instanceof AuthorizeCallback)) {
+        continue;
+      }
+
+      AuthorizeCallback ac = (AuthorizeCallback) callback;
+      String authenticationID = ac.getAuthenticationID();
+      String authorizationID = ac.getAuthorizationID();
+
+      if (StringUtils.isBlank(authenticationID) || StringUtils.isBlank(authorizationID)) {
+        LOG.debug("Missing authentication or authorization ID; skipping LDAP filter");
+        continue;
+      }
+
+      if (!authenticationID.equals(authorizationID)) {
+        LOG.debug("Skipping LDAP filter for mismatched auth IDs");
+        continue;
+      }
+
+      if (!enableLdapGroupCheck || filter == null) {
+        ac.setAuthorized(true);
+        continue;
+      }
+
+      String user = extractUserName(authenticationID);
+      boolean authorized = applyLdapFilter(user);
+      ac.setAuthorized(authorized);
```

Comment:
if `authorized` returned is false, we might need to set `ac.setAuthorized(false);` and remove the `AuthorizedID` from ac